inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,156 @@
+            const auto gatherIndicesShape = std::make_shared<ngraph::opset6::ShapeOf>(
+                matchedGatherIndices,
+                ngraph::as_type<ngraph::opset6::ShapeOf>(shapeOf)->get_output_type());
+            const auto axis = ngraph::as_type<ngraph::opset6::Gather>(matchedGather.get_node())->get_axis();
+
+            const auto gatherIndicesRank = matchedGatherIndices.get_partial_shape().rank().get_length();
+            const auto gatherDataRank = matchedGatherData.get_partial_shape().rank().get_length();
+            const auto gatherOutRank = gatherDataRank + gatherIndicesRank - 1;
+
+            std::vector<int64_t> squeezeOutIndices(gatherOutRank);
+            std::iota(squeezeOutIndices.begin(), squeezeOutIndices.end(), 0);
+
+            const auto normedSqueezeAxes = ngraph::normalize_axes(
+                matchedSqueezeAxes.get_node()->description(),
+                ngraph::as_type<ngraph::opset6::Constant>(matchedSqueezeAxes.get_node())->cast_vector<int64_t>(),
+                gatherOutRank);
+            const std::set<size_t, std::greater<size_t>> orderedSqueezeAxes(normedSqueezeAxes.begin(), normedSqueezeAxes.end());
+            for (const auto& squeezeAxis : orderedSqueezeAxes) {
+                squeezeOutIndices.erase(squeezeOutIndices.begin() + squeezeAxis);
+            }
+
+            std::vector<int64_t> transposeOutIndices(squeezeOutIndices.size());
+            auto transposePerm = ngraph::as_type<ngraph::opset6::Constant>(matchedTransposePerm.get_node())->cast_vector<int64_t>();
+            for (size_t i = 0; i < transposeOutIndices.size(); i++) {
+                transposeOutIndices[i] = squeezeOutIndices[transposePerm[i]];
+            }
+
+            ngraph::OutputVector gatherOutDims;
+            if (axis) {
+                gatherOutDims.push_back(gatherShapeElements(gatherDataShape, 0, axis));
+            }
+            if (matchedGatherIndices.get_partial_shape().rank().get_length()) {
+                gatherOutDims.push_back(gatherIndicesShape);
+            }
+            if (axis + 1 < gatherDataRank) {
+                gatherOutDims.push_back(gatherShapeElements(gatherDataShape, axis + 1, gatherDataRank - axis - 1));
+            }
+
+            const auto gatherOutShape = std::make_shared<ngraph::opset6::Concat>(gatherOutDims, 0);
+
+            const auto transposeOutShape = std::make_shared<ngraph::opset6::Gather>(
+                gatherOutShape,
+                ngraph::opset6::Constant::create(ngraph::element::i64, ngraph::Shape{transposeOutIndices.size()}, transposeOutIndices),
+                ngraph::opset6::Constant::create(ngraph::element::i64, ngraph::Shape{1}, {0}));
+
+            transposeOutShape->set_friendly_name(shapeOf->get_friendly_name());
+            shapeOf->get_default_output().replace(transposeOutShape);
+            wasGraphChanged = true;
+        }
+
+        m->clear_state();
```

Comment:
m->match already call `clear_state` method. So you don't need to do it manually.

Reply:
Removed.